### DISCUSSION TITLE
tree-sitter installation error (undefined exported symbol: "abort")

### DISCUSSION THREAD

**Discussion by brw** [Homebrew/discussions#1488] [Linux] [answered]
I'm not sure if this is homebrew-related at all but I thought I'd ask here anyway, maybe it has something to do with my environment. If it's unrelated I'll just open an issue in the tree-sitter repo.

On trying to install tree-sitter as a dependency of Neovim it fails on the `script/build-wasm` step with this error:
```
emcc: error: undefined exported symbol: "abort" [-Wundefined] [-Werror]
```
I'm not sure what to do about this, googling the error message does not seem to bring up anything and I can't find any related issues either.

gist-logs; https://gist.github.com/71690df64796b675a7abfd6bd9c66c7d

**Comment by carlocab** [comment, marked as answer]
I'd try replicating the steps that the formula does outside of `brew`. If you encounter the same error without using `brew`, that's an upstream problem and should be reported there.

**Reply by brw** [reply]
Yea it looks like it does. Probably not homebrew environment-related either since it also happens when using the `--docker` flag with `build-wasm`.

I honestly don't know much about emscripten/wasm/tree-sitter since I just want to use Neovim but I'll just ask in the tree-sitter repo then I suppose.

**Reply by carlocab** [reply]
If you just want to use `tree-sitter` for Neovim, you can probably do with building `tree-sitter` with this patch applied to the formula:

```diff
diff --git a/Formula/tree-sitter.rb b/Formula/tree-sitter.rb
index c736c94f81..0a093f01a7 100644
--- a/Formula/tree-sitter.rb
+++ b/Formula/tree-sitter.rb
@@ -22,22 +22,6 @@ class TreeSitter < Formula
   def install
     system "make"
     system "make", "install", "PREFIX=#{prefix}"
-
-    # NOTE: This step needs to be done *before* `cargo install`
-    cd "lib/binding_web" do
-      system "npm", "install", *Language::Node.local_npm_install_args
-    end
-    system "script/build-wasm"
-
-    cd "cli" do
-      system "cargo", "install", *std_cargo_args
-    end
-
-    # Install the wasm module into the prefix.
-    # NOTE: This step needs to be done *after* `cargo install`.
-    %w[tree-sitter.js tree-sitter-web.d.ts tree-sitter.wasm package.json].each do |file|
-      (lib/"binding_web").install "lib/binding_web/#{file}"
-    end
   end
 
   test do
```
You can also remove the `node`, `emscripten`, and `rust` dependencies. If you still need the `cli`, you can keep the three lines around `cargo install`, but you'll still need `rust` and (I think) `node`.

**Reply by brw** [reply]
Actually I just found the solution. It seems that emscripten >=2.0.18 is not backwards compatible with <=2.0.17. 2.0.17 is the latest that works with tree-sitter wasm according to this comment: https://github.com/tree-sitter/tree-sitter/issues/1098#issuecomment-842326203.

I've also verified that `script/build-wasm` runs fine with emcc 2.0.17. Should the emscripten dependency for tree-sitter be pointed at 2.0.17 instead of 2.0.20?

**Reply by brw** [reply]
> If you just want to use `tree-sitter` for Neovim, you can probably do with building `tree-sitter` with this patch applied to the formula:
> 
> ```diff
> diff --git a/Formula/tree-sitter.rb b/Formula/tree-sitter.rb
> index c736c94f81..0a093f01a7 100644
> --- a/Formula/tree-sitter.rb
> +++ b/Formula/tree-sitter.rb
> @@ -22,22 +22,6 @@ class TreeSitter < Formula
>    def install
>      system "make"
>      system "make", "install", "PREFIX=#{prefix}"
> -
> -    # NOTE: This step needs to be done *before* `cargo install`
> -    cd "lib/binding_web" do
> -      system "npm", "install", *Language::Node.local_npm_install_args
> -    end
> -    system "script/build-wasm"
> -
> -    cd "cli" do
> -      system "cargo", "install", *std_cargo_args
> -    end
> -
> -    # Install the wasm module into the prefix.
> -    # NOTE: This step needs to be done *after* `cargo install`.
> -    %w[tree-sitter.js tree-sitter-web.d.ts tree-sitter.wasm package.json].each do |file|
> -      (lib/"binding_web").install "lib/binding_web/#{file}"
> -    end
>    end
>  
>    test do
> ```
> 
> You can also remove the `node`, `emscripten`, and `rust` dependencies. If you still need the `cli`, you can keep the three lines around `cargo install`, but you'll still need `rust` and (I think) `node`.

Ah I didn't see your reply yet while writing mine. Thanks for the patch and the explanation! I didn't exactly know if I needed everything or not. That'll probably work.

**Reply by carlocab** [reply]
> Should the emscripten dependency for tree-sitter be pointed at 2.0.17 instead of 2.0.20?

Homebrew doesn't package 2.0.17. We might have to at the next `tree-sitter` version bump, but until then, let's leave it as-is.

**Reply by brw** [reply]
Ah I see, alright then. Thanks for your help so far!

**Comment by carlocab** [comment]
`tree-sitter` 0.19.5 appears to be compatible with the newest `emscripten`: https://github.com/Homebrew/homebrew-core/pull/77711

So your local workaround for `tree-sitter` should hopefully no longer be necessary soon.

I've also made sure that future `emscripten` CI runs include a test for building `tree-sitter` from source (https://github.com/Homebrew/homebrew-test-bot/pull/613), so we should be far less likely to end up in a situation where `tree-sitter` can't actually be built with the version of `emscripten` shipped by Homebrew.

**Reply by brw** [reply]
That's awesome, thanks a lot!
